numba/array_analysis.py

```diff
@@ -41,846 +52,1266 @@
                 random_3arg_sizelast +
                 ['random.randint', 'random.triangular'])
 
+@intrinsic
+def assert_equiv(typingctx, *val):
+    """
+    A function that asserts the inputs are of equivalent size,
+    and throws runtime error when they are not. The input is
+    a vararg that contains an error message, followed by a set
+    of objects of either array, tuple or integer.
+    """
+    # Arguments must be either array, tuple, or integer
+    assert all(map(lambda a: (isinstance(a, types.ArrayCompatible) or
+                             isinstance(a, types.BaseTuple) or
+                             isinstance(a, types.Integer)), val))
+
+    def codegen(context, builder, sig, args):
+        assert(len(args) == 1) # it is a vararg tuple
+        tup = cgutils.unpack_tuple(builder, args[0])
+        tup_type = sig.args[0]
+        msg = sig.args[0][0].value
+
+        def unpack_shapes(a, aty):
+            if isinstance(aty, types.ArrayCompatible):
+                ary = context.make_array(aty)(context, builder, a)
+                return cgutils.unpack_tuple(builder, ary.shape)
+            elif isinstance(aty, types.BaseTuple):
+                return cgutils.unpack_tuple(builder, a)
+            else: # otherwise it is a single integer
+                return [a]
+
+        def pairwise(a, aty, b, bty):
+            ashapes = unpack_shapes(a, aty)
+            bshapes = unpack_shapes(b, bty)
+            assert len(ashapes) == len(bshapes)
+            for (m, n) in zip(ashapes, bshapes):
+                m_eq_n = builder.icmp(lc.ICMP_EQ, m, n)
+                with builder.if_else(m_eq_n) as (then, orelse):
+                    with then:
+                        pass
+                    with orelse:
+                        context.call_conv.return_user_exc(
+                            builder, AssertionError, (msg,))
+
+        for i in range(1, len(tup_type) - 1):
+            pairwise(tup[i], tup_type[i], tup[i+1], tup_type[i+1])
+        r = context.get_constant_generic(builder, types.NoneType, None)
+        return r
+    return signature(types.none, *val), codegen
+
+
+class EquivSet(object):
+    """EquivSet keeps track of equivalence relations between
+    a set of objects.
+    """
+    def __init__(self, obj_to_ind = None, ind_to_obj = None, next_ind = 0):
```

Comment:
The keyword arguments are for internal use only (by clone). Will make a docstring.